beacon-chain/blockchain/core.go

```diff
@@ -319,59 +303,105 @@ func (b *BeaconChain) saveBlock(block *types.Block) error {
-		msg := make([]byte, binary.MaxVarintLen64)
-		var signedHashesStr []byte
-		for _, parentHash := range parentHashes {
-			signedHashesStr = append(signedHashesStr, parentHash.Bytes()...)
-			signedHashesStr = append(signedHashesStr, byte(' '))
-		}
-		binary.PutUvarint(msg, attestation.Slot%params.CycleLength)
-		msg = append(msg, signedHashesStr...)
-		binary.PutUvarint(msg, attestation.ShardId)
-		msg = append(msg, attestation.ShardBlockHash...)
+	// Hash parentHashes + shardID + slotNumber + shardBlockHash into a message to use to
+	// to verify with aggregated public key and aggregated attestation signature.
+	msg := make([]byte, binary.MaxVarintLen64)
+	var signedHashesStr []byte
+	for _, parentHash := range parentHashes {
+		signedHashesStr = append(signedHashesStr, parentHash.Bytes()...)
+		signedHashesStr = append(signedHashesStr, byte(' '))
+	}
+	binary.PutUvarint(msg, attestation.Slot%params.CycleLength)
+	msg = append(msg, signedHashesStr...)
+	binary.PutUvarint(msg, attestation.ShardId)
+	msg = append(msg, attestation.ShardBlockHash...)
 
-		msgHash := blake2b.Sum512(msg)
+	msgHash := blake2b.Sum512(msg)
 
-		log.Debugf("Attestation message for shard: %v, slot %v, block hash %v is: %v",
-			attestation.ShardId, attestation.Slot, attestation.ShardBlockHash, msgHash)
+	log.Debugf("Attestation message for shard: %v, slot %v, block hash %v is: %v",
+		attestation.ShardId, attestation.Slot, attestation.ShardBlockHash, msgHash)
 
-		// TODO: Verify msgHash against aggregated pub key and aggregated signature.
-	}
+	// TODO: Verify msgHash against aggregated pub key and aggregated signature.
 	return nil
 }
 
+// calculateBlockVoteCache calculates and updates active state's block vote cache.
+func (b *BeaconChain) calculateBlockVoteCache(attestationIndex int, block *types.Block) (map[*common.Hash]*types.VoteCache, error) {
+	attestation := block.Attestations()[attestationIndex]
+	newVoteCache := b.ActiveState().GetBlockVoteCache()
+	parentHashes := b.getSignedParentHashes(block, attestation)
+	attesterIndices, err := b.getAttesterIndices(attestation)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, h := range parentHashes {
+		// Skip calculating for this hash if the hash is part of oblique parent hashes.
+		var skip bool
+		for _, obliqueParentHash := range attestation.ObliqueParentHashes {
+			if bytes.Equal(h.Bytes(), obliqueParentHash) {
+				skip = true
+			}
+		}
+		if skip {
+			continue
+		}
+
+		// Initialize vote cache of a given block hash if it doesn't exist already.
+		if !b.ActiveState().IsVoteCacheEmpty(h) {
+			newVoteCache[h] = &types.VoteCache{VoterIndices: []uint32{}, VoteTotalDeposit: 0}
+		}
+
+		// Loop through attester indices, if the attester has voted but was not accounted for
+		// in the cache, then we add attester's index and balance to the block cache.
+		for i, attesterIndex := range attesterIndices {
+			var newAttester bool
+			if utils.CheckBit(attestation.AttesterBitfield, i) {
```

Comment:
nitpick, but I prefer exiting early for better readability (less nesting). So
```
if !utils.CheckBit(attestation.AttesterBitfield, i) {
    continue
}
...
```

Reply:
Good point. Changed